File: simulation/swarm/missiles/hydra_70.cc
#include "simulation/swarm/missiles/hydra_70.h"

#include <Eigen/Dense>
#include <memory>
#include <stdexcept>
#include <utility>
#include <vector>

#include "absl/strings/str_format.h"
#include "simulation/swarm/agent.h"
#include "simulation/swarm/missiles/micromissile.h"
#include "simulation/swarm/proto/agent.pb.h"

namespace swarm::missile {

std::vector<std::unique_ptr<agent::Agent>> Hydra70::Spawn(const double t) {
  if (has_spawned_) {
    return std::vector<std::unique_ptr<agent::Agent>>();
  }

  const auto num_submunitions = submunitions_config().num_submunitions();
  const auto launch_time = dynamic_config().launch_config().launch_time();
  const auto submunitions_launch_time =
      submunitions_config().launch_config().launch_time();
  if (t >= t_creation_ + launch_time + submunitions_launch_time) {
    // Define the missile configuration for the submunitions.
    const auto submunitions_type =
        submunitions_config().agent_config().missile_type();
    AgentConfig submunitions_missile_config(
        submunitions_config().agent_config());
    submunitions_missile_config.mutable_initial_state()->CopyFrom(state());

    // Create the missiles for the submunitions.
    std::vector<std::unique_ptr<agent::Agent>> spawned_missiles;
    spawned_missiles.reserve(num_submunitions);
    for (int i = 0; i < num_submunitions; ++i) {
      spawned_missiles.emplace_back(CreateSubmunition(
          submunitions_type, submunitions_missile_config, t, /*ready=*/true));
    }
    has_spawned_ = true;
    return spawned_missiles;
  }
  return std::vector<std::unique_ptr<agent::Agent>>();
}

void Hydra70::UpdateMidCourse(const double t) {
  // The Hydra-70 rocket is unguided, so only consider gravity and drag.
  Eigen::Vector3d acceleration_input = Eigen::Vector3d::Zero();

  // Calculate and set the total acceleration.
  const auto acceleration = CalculateAcceleration(acceleration_input);
  state_.mutable_acceleration()->set_x(acceleration(0));
  state_.mutable_acceleration()->set_y(acceleration(1));
  state_.mutable_acceleration()->set_z(acceleration(2));
}

template <typename... Args>
std::unique_ptr<Missile> Hydra70::CreateSubmunition(const MissileType type,
                                                    Args&&... args) {
  switch (type) {
    case MissileType::MICROMISSILE: {
      return std::make_unique<Micromissile>(std::forward<Args>(args)...);
    }
    default: {
      throw std::invalid_argument(
          absl::StrFormat("Invalid submunition type: %d.", type));
    }
  }
}

}  // namespace swarm::missile

File: simulation/swarm/missiles/py/hydra_70.py
"""The Hydra-70 class represents the dynamics of a single unguided Hydra-70 rocket."""

import google.protobuf
import numpy as np
from simulation.swarm.proto.agent_pb2 import AgentConfig, MissileType
from simulation.swarm.proto.static_config_pb2 import StaticConfig

from simulation.swarm.missiles.py.micromissile import Micromissile
from simulation.swarm.missiles.py.missile_interface import Missile
from simulation.swarm.py import constants

# Hydra-70 submunitions type enumeration to the missile class.
HYDRA_70_SUBMUNITIONS_TYPE_ENUM_TO_CLASS = {
    MissileType.MICROMISSILE: Micromissile,
}


class Hydra70(Missile):
    """Hydra-70 dynamics.

    Attributes:
        has_spawned: A boolean indicating whether the rocket has spawned submunitions.
    """

    def __init__(
        self,
        missile_config: AgentConfig,
        ready: bool = True,
        t_creation: float = 0,
    ) -> None:
        super().__init__(missile_config, ready, t_creation)
        self.has_spawned = False

    @property
    def static_config(self) -> StaticConfig:
        """Returns the static configuration of the Hydra-70 rocket."""
        static_config_file_path = (
            "simulation/swarm/configs/missiles/hydra_70.pbtxt")
        with open(static_config_file_path, "r") as static_config_file:
            static_config = google.protobuf.text_format.Parse(
                static_config_file.read(), StaticConfig())
        return static_config

    def assignable_to_target(self) -> bool:
        """Returns whether a target can be assigned to the missile."""
        return False

    def spawn(self, t: float) -> list[Missile]:
        """Spawns new missiles.

        Args:
            t: Time in seconds.

        Returns:
            A list of newly spawned missiles.
        """
        if self.has_spawned:
            return []

        num_submunitions = self.submunitions_config.num_submunitions
        launch_time = self.dynamic_config.launch_config.launch_time
        submunitions_launch_time = (
            self.submunitions_config.launch_config.launch_time)
        if t >= self.t_creation + launch_time + submunitions_launch_time:
            # Define the missile configuration for the submunitions.
            submunitions_type = self.submunitions_config.agent_config.missile_type
            submunitions_missile_config = AgentConfig()
            submunitions_missile_config.CopyFrom(
                self.submunitions_config.agent_config)
            submunitions_missile_config.initial_state.CopyFrom(self.state)

            # Create the missiles for the submunitions.
            spawned_missiles = [
                HYDRA_70_SUBMUNITIONS_TYPE_ENUM_TO_CLASS[submunitions_type](
                    submunitions_missile_config, t_creation=t)
                for _ in range(num_submunitions)
            ]
            self.has_spawned = True
            return spawned_missiles
        return []

    def _update(self, t: float) -> None:
        """Updates the agent's state in the midcourse and terminal flight
        phase.

        Args:
            t: Time in seconds.
        """
        # The Hydra-70 rocket is unguided, so only consider gravity and drag.
        acceleration_input = np.zeros(3)
        acceleration = self._calculate_total_acceleration(acceleration_input)
        (
            self.state.acceleration.x,
            self.state.acceleration.y,
            self.state.acceleration.z,
        ) = acceleration
